pkg/controllers/counter/controller.go

```diff
@@ -0,0 +1,127 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package counter
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha5"
+)
+
+// Controller for the resource
+type Controller struct {
+	ctx        context.Context
+	kubeClient client.Client
+}
+
+// NewController is a constructor
+func NewController(ctx context.Context, kubeClient client.Client) *Controller {
+	return &Controller{
+		ctx:        ctx,
+		kubeClient: kubeClient,
+	}
+}
+
+// Reconcile a control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logging.FromContext(ctx).Infof("Updating resource counts on provisioner")
+	// Retrieve the provisioner
+	provisioner := &v1alpha5.Provisioner{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, provisioner); err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+	persisted := provisioner.DeepCopy()
+
+	// Determine resource usage and update provisioner.status.resources
+	resourceCounts, err := c.resourceCountsFor(ctx, provisioner.Name)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	provisioner.Status.Resources = resourceCounts
+	if err := c.kubeClient.Status().Patch(ctx, provisioner, client.MergeFrom(persisted)); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to persist changes to %s, %w", req.NamespacedName, err)
+	}
+
+	// Refresh the reconciler state values every 5 minutes irrespective of node events
+	return reconcile.Result{RequeueAfter: 5 * time.Minute}, nil
+}
+
+func (c *Controller) resourceCountsFor(ctx context.Context, provisionerName string) (v1.ResourceList, error) {
+	nodes := v1.NodeList{}
+	if err := c.kubeClient.List(ctx, &nodes, client.MatchingLabels{v1alpha5.ProvisionerNameLabelKey: provisionerName}); err != nil {
+		return nil, err
+	}
+
+	var cpu = resource.NewScaledQuantity(0, 0)
+	var memory = resource.NewScaledQuantity(0, resource.Giga)
+
+	for _, node := range nodes.Items {
```

Comment:
Can we factor in GPUs here?

Reply:
GPUs is a little more work. Node objects don't have any notion of GPUs as far as I can tell so I've to go the long route of fetching the instanceType from the node labels and then asking the cloudprovider about the GPUs that instanceType comes with. 

I was thinking of building a separate cache for all instanceTypes (I think there was a cloudprovider interface that gets us that data already) in this controller.

Reply:
Basically, I'm happy to do it but was thinking I'll do it either in my next PR or after the CPU memory bit looks good :)